p2p/src/peer_manager/mod.rs

```diff
@@ -313,26 +350,36 @@ where
         // accepted even if it's valid. The peer is still reported to the PeerDb which
         // knows of all peers and later on if the number of connections falls below
         // the desired threshold, `PeerManager::heartbeat()` may connect to this peer.
-        if self.peerdb.active_peer_count() >= MAX_ACTIVE_CONNECTIONS {
+        if self.active_peer_count() >= MAX_ACTIVE_CONNECTIONS {
             return Err(P2pError::PeerError(PeerError::TooManyPeers));
         }
 
         self.accept_connection(address, Role::Inbound, info, receiver_address).await
     }
 
-    /// Close connection to a remote peer
+    /// The connection to a remote peer is reported as closed.
     ///
-    /// The decision to close the connection is made either by the user via RPC
-    /// or by the [`PeerManager::heartbeat()`] function which has decided to cull
-    /// this connection in favor of another potential connection.
-    fn close_connection(&mut self, peer_id: T::PeerId) -> crate::Result<()> {
+    /// This can happen when the remote peer has dropped its connection
+    /// or if a disconnect request has been sent by PeerManager to the backend.
+    fn connection_closed(&mut self, peer_id: T::PeerId) -> crate::Result<()> {
         // The backend is always sending ConnectionClosed event when somebody disconnects, ensure that the peer is active
-        if self.peerdb.is_active_peer(&peer_id) {
-            log::debug!("connection closed for peer {peer_id}");
+        if self.is_active_peer(&peer_id) {
```

Comment:
I'm not sure I understand the specific reason for this logic. You check if a peer exists, then remove it, then expect if it's not there... why not do something like `if let Some(peer) = self.peers.remove(peer_id)`? This looks ugly due to the statefulness of it, but then this is what's practically being done. This should be cleaned up a little, I think.

Reply:
Sure, updated, thanks.